File: runtime/android/xwalk_view.cc
// Copyright (c) 2013 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "xwalk/runtime/android/xwalk_view.h"

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "base/android/scoped_java_ref.h"
#include "base/bind.h"
#include "base/lazy_instance.h"
#include "content/shell/shell.h"
#include "content/shell/shell_browser_context.h"
#include "content/shell/shell_content_browser_client.h"
#include "content/public/browser/web_contents.h"
#include "jni/XWalkView_jni.h"

using base::android::ScopedJavaLocalRef;

namespace {

struct GlobalState {
  GlobalState() {}
  base::android::ScopedJavaGlobalRef<jobject> j_shell_manager;
};

base::LazyInstance<GlobalState> g_global_state = LAZY_INSTANCE_INITIALIZER;

}  // namespace

namespace xwalk {

bool RegisterXWalkView(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

void InitWithWebContents(content::WebContents *web_contents) {
  JNIEnv *env = base::android::AttachCurrentThread();
  Java_XWalkView_initWithWebContents(env, g_global_state.Get().j_shell_manager.obj(), reinterpret_cast<jint>(web_contents));
}

static void Init(JNIEnv* env, jclass clazz, jobject obj) {
  g_global_state.Get().j_shell_manager.Reset(
      base::android::ScopedJavaLocalRef<jobject>(env, obj));
}

}   // namespace xwalk

File: runtime/android/java/src/com/intel/xwalk/XWalkView.java
// Copyright (c) 2013 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.intel.xwalk;

import android.content.Context;
import android.util.AttributeSet;
import android.webkit.WebView;
import android.widget.FrameLayout;

import org.chromium.base.CalledByNative;
import org.chromium.base.JNINamespace;
import org.chromium.content.browser.ContentView;
import org.chromium.content.browser.ContentViewRenderView;
import org.chromium.content.browser.LoadUrlParams;
import org.chromium.ui.WindowAndroid;

@JNINamespace("xwalk")
public class XWalkView extends FrameLayout {
   private WindowAndroid mWindowAndroid;
   private ContentView mContentView;
   private ContentViewRenderView mContentViewRenderView; // SurfaceView

   public XWalkView(Context context, AttributeSet attrs) {
        super(context, attrs);
        nativeInit(this);

        mContentViewRenderView = new ContentViewRenderView(context) {
            @Override
            protected void onReadyToRender() {
                mContentView.loadUrl(new LoadUrlParams("http://www.google.com"));
                mContentView.clearFocus();
                mContentView.requestFocus();
            }
        };
    }

    public void setWindow(WindowAndroid window) {
        mWindowAndroid = window;
    }

    public void loadUrl(String url) {
    }

    @SuppressWarnings("unused")
    @CalledByNative
    private void initWithWebContents(int webContents) {
        addView(mContentViewRenderView, new FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT));

        mContentView = ContentView.newInstance(getContext(), webContents,
            mWindowAndroid, ContentView.PERSONALITY_VIEW);

        addView(mContentView, new FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT));

        mContentViewRenderView.setCurrentContentView(mContentView);
    }

    private static native void nativeInit(Object view);
}

